### DISCUSSION TITLE
[PSD18] Esercizi Slide 33

### DISCUSSION THREAD

**Discussion by matypist** [sapienzastudentsnetwork/psd2223#191] [Esercizi (A-L)]
Costruire una memoria di **1 MB** e...

1. larghezza di parola di **32** bits a partire da blocchi di memoria di dimensioni $2^{18} * 8$

2. larghezza di parola di **32** bits a partire da blocchi di memoria di dimensioni $2^{16} * 32$

3. larghezza di parola di **8** bits a partire da blocchi di memoria di dimensioni $2^{18} * 32$

**Comment by Elia-Belli** [comment]
- **1)** Un blocco di memoria di dimensioni $2^{18}*8$ ha $2^{21}\ bits$ = $2\ Mbit$ = $256\ KB$, quindi per realizzare una memoria di $1\ MB$ ne servono $4$.
Inoltre, siccome la memoria da realizzare ha larghezza di parola di $32$ bits ed i blocchi solo di $8\ bit$, divideremo i $32\ bits$ in $4$ porzioni da $8$, ognuna memorizzata in uno dei blocchi.
![M1](https://user-images.githubusercontent.com/107944476/212557673-644b9176-55c5-4a3d-96c7-12717e522f71.jpg)

- **2)** Un blocco di memoria di dimensioni $2^{16}*32$ ha $2^{21}\ bits$ = $2\ Mbit$ = $256\ KB$, quindi per realizzare una memoria di $1\ MB$ ne servono $4$.
Ogni informazione di $32\ bits$ verrà memorizzata unicamente in un blocco a differenza dell'esercizio precedente, quindi dobbiamo stabilire una codifica per decidere in quale blocco inviare l'informazione, che è totalmente arbitraria, in questo caso utilizzerò i 2 bit più significativi dell' $ADDRESS$: se $ADDR[15:14] = 00$ l'informazione sarà memorizzata in $M0$, se $01$ in $M1$, se $10$ in $M2$ e se $11$ in $M3$, in questo caso un $DEC[2:4]$ ed un $MUX[4:1]$ sono gli strumentopoli giusti.
![M2](https://user-images.githubusercontent.com/107944476/212862057-701f25bb-a027-46c4-8fea-267926bcd17a.jpg)
**Spiegazione Circuito**: in questa memoria il meccanismo di lettura e scrittura sono separati, abbiamo 3 input $WDATA$ che fornisce l'informazione da memorizzare, $ADDR$ che indica l'indirizzo su cui leggere o scrivere, $WE$ che indica se leggiamo o scriviamo e l'output $RDATA$ che restituisce l'informazione letta. 
Quando si legge si pesca l'informazione da tutte le memorie $M_i$ ma tramite un Mux solo quella giusta è scelta (in realtà ci sono 32 mux, uno per bit, ma per comodità ne ho disegnato uno).
Quando si scrive, l'Enabled Decoder è attivato da $WE$ (quando $WE = 0$ tutti gli out sono 0), i bit più significativi dell' $ADDR$ attivano il $WE$ del blocco su cui scrivere, $ADDR$ indica l'indirizzo in cui scrivere e $WDATA$ fornisce l'informazione.

- **3)** Un blocco di memoria di dimensioni $2^{18}*32$ ha $2^{23}\ bits$ = $8\ Mbit$ = $1\ MB$, quindi ne basta una. Siccome il blocco ha larghezza di parola $32$ ma è richiesta di $8$ dovremo applicare una codifica per leggere e scrivere solo la porzione di "riga" che ci interessa, siccome una riga di $32\ bits$ può contenere $4$ informazione da $8\ bits$ ci bastano 2 bit per la codifica e possiamo usare la stessa dell'esercizio precedente per comodità.
**Attenzione**: non è possibile leggere o scrivere soltanto una porzione di una word, pertanto quando leggiamo dovremo selezionare solo i bit che ci interessano e quando scriviamo dovremo riscrivere tutta la word modificando solo 8 bits.
**Lettura**: selezioniamo la word tramite l' $ADDR$, in base ai 2 bit più significativi sappiamo quale porzione leggere ( $00$ la prima da sinistra, $01$ la seconda, $10$ la terza, $11$ la quarta) utilizziamo quindi un Mux per selezionare i bit desiderati (32 mux per la precisione).
**Scrittura**: prima di scrivere dobbiamo copiare la word all'indirizzo, potremmo salvarla in un registro e modificare gli 8 bit da scrivere, poi abilitiamo il segnale $WE$ e riscriviamo l'intera riga, con il risultato che solo i bit modificati saranno differenti. La scrittura sarà più lenta della scrittura richiedendo più operazioni e quindi più colpi di clock.
**Circuito**: se volete realizzare il circuito potete seguire queste direttive o proporre un metodo più efficiente.

**Comment by fraxb04** [comment]
@Elia-Belli 
non ho ben capito nel secondo punto, la parte del decoder... cioè il decoder non ha dei selettori, ma solamente degli input, ed attiva la codifica one-hot... per come è disegnato sembra che abbia, oltre ai due input (i due Address), anche un selettore rappresentato dal write enable.

**Comment by Elia-Belli** [comment]
> @Elia-Belli non ho ben capito nel secondo punto, la parte del decoder... cioè il decoder non ha dei selettori, ma solamente degli input, ed attiva la codifica one-hot... per come è disegnato sembra che abbia, oltre ai due input (i due Address), anche un selettore rappresentato dal write enable.

Infatti non è un decoder classico, ma un Enabled Decoder.

**Comment by fraxb04** [comment]
@Elia-Belli, ok ma dove posso trovarlo nella teoria (ad esempio sugli appunti di Exyss)

**Comment by fraxb04** [comment]
@Elia-Belli 
corretto

**Comment by CiottoloMaggico** [comment]
![immagine](https://user-images.githubusercontent.com/79638284/219012227-56c6a1c0-6df2-4fe2-9621-e90d3efa631d.png)
![immagine](https://user-images.githubusercontent.com/79638284/219012274-d967b6c1-9d54-435f-ba62-c887654040b9.png)
![immagine](https://user-images.githubusercontent.com/79638284/219012302-106071fd-e60b-4dbb-b714-a39bd841c9ea.png)
